### PR TITLE
feat: factorize  docker build/push on tests

### PR DESCRIPTION
**Pull request by spencerdrak**
PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

**Review by cgdolan — APPROVED** [review]

File: .github/workflows/build-test-docker.yaml [hunk 1]
```diff
@@ -2,26 +2,85 @@ name: build-test-docker
...
+    inputs:
+      docker-tags:
+        required: true
+        type: string
+        description: A multi-line string in the format accepted by docker metadata tag action for the tags to apply to the image
+      artifact-name:
+        required: true
+        type: string
+        description: Name (key) to use when uploading the docker image tarball as a artifact
+      repository-name:
```
Comment (aryx) [inline .github/workflows/build-test-docker.yaml:14]: Why do we need this thing as input? Isn't it always returntocorp/semgrep?

File: .github/workflows/build-test-docker.yaml [hunk 2]
```diff
@@ -2,26 +2,85 @@ name: build-test-docker
...
+        required: true
+        type: string
+        description: The repository/name of the docker image to push, e.g., returntocorp/semgrep
+      file:
```
Comment (aryx) [inline .github/workflows/build-test-docker.yaml:18]: Same, isn't it always semgrep/Dockerfile?

File: .github/workflows/build-test-docker.yaml [hunk 3]
```diff
@@ -2,26 +2,85 @@ name: build-test-docker
...
-  release-docker:
+  build-test-docker:
     name: Build and test Semgrep Docker image
     runs-on: ubuntu-latest
     steps:
       - name: Set up Docker Buildx
         uses: docker/setup-buildx-action@v2
-
+      - id: meta
+        name: Set tags and labels
```
Comment (aryx) [inline .github/workflows/build-test-docker.yaml:58]: This is mystery to me. Why do we need that?

File: .github/workflows/build-test-docker.yaml [hunk 4]
```diff
@@ -2,26 +2,85 @@ name: build-test-docker
...
+      - name: Load image
+        run: |
+          docker load --input /tmp/image.tar
       - name: Checkout
         uses: actions/checkout@v3
-
-      - name: Check the semgrep Docker image
-        run: ./scripts/validate-docker-build.sh candidate-image
+      - name: Test Image
+        if: ${{ inputs.enable-tests }}
```
Comment (aryx) [inline .github/workflows/build-test-docker.yaml:80]: Why do we want to make this an option? Don't we always want to test?

File: .github/workflows/push-docker.yaml [hunk 5]
```diff
@@ -0,0 +1,46 @@
+name: push-docker
+
+on:
+  workflow_dispatch:
+    inputs:
+      artifact-name:
+        required: true
+        type: string
+        description: Name (key) to use when uploading the docker image tarball as a artifact
+      repository-name:
```
Comment (aryx) [inline .github/workflows/push-docker.yaml:10]: Same, why make it an input variable if it's fixed and it's a constant?

File: .github/workflows/build-test-docker.yaml [hunk 6]
```diff
@@ -2,26 +2,85 @@ name: build-test-docker
 
 on:
   workflow_dispatch:
+    inputs:
```
Comment (aryx) [inline .github/workflows/build-test-docker.yaml:5]: Please add a toplevel comment explaining the goal of this workflow (you can see at the other workflow in the same directory for example).

File: .github/workflows/push-docker.yaml [hunk 7]
```diff
@@ -0,0 +1,46 @@
+name: push-docker
```
Comment (aryx) [inline .github/workflows/push-docker.yaml:1]: Please, add a toplevel comment.

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -182,6 +182,14 @@ jobs:
   build-test-docker:
     uses: ./.github/workflows/build-test-docker.yaml
     secrets: inherit
+    with:
```
Comment (aryx) [inline .github/workflows/release.yml:185]: Oh ok I think I understand now.
So maybe you can add in the build-test-docker.yaml toplevel comment something like
"This is mostly independent of semgrep. The goal is to factorize xxx?
This is mostly a small wrapper around docker/setup-buildx-action@v2 and docker/metadata-action@v4".

Even though you mention hardcoded path like scripts/validate-docker in it ...

**Review by aryx — COMMENTED** [review]
It says factorize but your PR is actually adding lots of code ...
+167 −80

**Pull request merged**